experimental/dds/tree2/src/feature-libraries/simple-tree/proxyNodeCursor.ts

```diff
@@ -0,0 +1,193 @@
+ */
+
+import { assert } from "@fluidframework/core-utils";
+
+import {
+	EmptyKey,
+	rootFieldKey,
+	type FieldKey,
+	type TreeTypeSet,
+	type Value,
+	CursorLocationType,
+} from "../../core";
+// eslint-disable-next-line import/no-internal-modules
+import { leaf } from "../../domains/leafDomain";
+import {
+	type ContextuallyTypedNodeData,
+	type TreeDataContext,
+	getPossibleTypes,
+	isFluidHandle,
+	isPrimitiveValue,
+} from "../contextuallyTyped";
+import {
+	type CursorAdapter,
+	type CursorWithNode,
+	stackTreeNodeCursor,
+	stackTreeFieldCursor,
+} from "../treeCursorUtils";
+import { TreeNodeSchema } from "../typed-schema";
+import { type ProxyNode } from "./types";
+/**
+ * Creates a {@link CursorAdapter} capable of mapping proxy-domain trees to the editable tree domain.
+ *
+ * @param context - The {@link TreeEntity.context} of the parent node under which the tree is being inserted.
+ * @param typeSet - The types allowed under the parent node under which the tree is being inserted.
+ *
+ * @privateRemarks
+ * This is exported solely for testing purposes.
+ * Consumers should use {@link cursorFromProxyTree} instead.
+ */
+export function createProxyTreeAdapter<TNode extends ProxyNode<TreeNodeSchema, "javaScript">>(
+	context: TreeDataContext,
+	typeSet: TreeTypeSet,
+): CursorAdapter<TNode> {
+	return {
+		value: (node: TNode) => {
+			return isPrimitiveValue(node) || node === null || isFluidHandle(node)
+				? (node as Value)
+				: undefined;
+		},
+		type: (node: TNode) => {
+			if (node === undefined) {
+				throw new Error(
+					"Cursor adapter was passed an undefined node. This is unsupported.",
+				);
+			}
+			if (node === null) {
+				return leaf.null.name;
+			}
+
+			switch (typeof node) {
+				case "number":
+					return leaf.number.name;
+				case "string":
+					return leaf.string.name;
+				case "boolean":
+					return leaf.boolean.name;
+				default: {
+					if (isFluidHandle(node)) {
+						return leaf.handle.name;
+					}
+
+					const possibleTypes = getPossibleTypes(
+						context,
+						typeSet,
+						node as ContextuallyTypedNodeData,
+					);
+					assert(
+						possibleTypes.length !== 0,
+						"data is incompatible with all types allowed by the schema",
+					);
+					assert(
+						possibleTypes.length === 1,
+						"data is compatible with more than one type allowed by the schema",
+					);
+					return possibleTypes[0];
+				}
+			}
+		},
+		keysFromNode: (node: TNode) => {
+			if (node === undefined) {
+				throw new Error(
+					"Cursor adapter was passed an undefined node. This is unsupported.",
+				);
+			}
+			switch (typeof node) {
+				case "object":
+					if (node === null) {
+						return [];
+					} else if (Array.isArray(node)) {
+						// A list with no children is considered to not have any fields.
+						// If it has children, return the special symbol indicating a list.
+						return node.length === 0 ? [] : [EmptyKey];
+					} else if (node instanceof Map) {
```

Comment:
Are TreeMapNodes nodes actually instances of map?

Reply:
No, but the object that you use in inserted content to specify a new Map is. These aren't `TreeMapNodes`, they're just POJOs with Arrays and Maps in them sometimes.

Reply:
`SharedTreeMap` (the proxy type) is typed as a Map, yes. Assuming that's what you're referring to.

Reply:
I believe the intention is to allow other forms of input in the future, but for now we strictly take TypeScript Maps.